rust/xaynet-server/src/storage/redis.rs

```diff
@@ -292,22 +298,48 @@ impl Connection {
     ///
     /// The score/counter of the given mask is incremented by `1`.
     /// The maximum length of a serialized mask is 512 Megabytes.
-    pub async fn incr_mask_count(mut self, mask: &MaskMany) -> RedisResult<()> {
+    pub async fn incr_mask_count(
+        mut self,
+        sum_pk: &SumParticipantPublicKey,
+        mask: &MaskObject,
+    ) -> RedisResult<MaskDictIncr> {
         debug!("increment mask count");
-        // https://redis.io/commands/zincrby
-        // > Return value
-        //   Bulk string reply: the new score of member (a double precision floating point number),
-        //   represented as string.
-        //
-        // We ignore the return value because we are not interested in it. We will use the method
-        // `get_best_masks` instead.
-        self.connection
-            .zincr("mask_dict", MaskObjectWrite::from(mask), 1_usize)
+        let script = Script::new(
+            r#"
+                -- lua lists (tables) start at 1
+                local sum_pk = ARGV[1]
+
+                -- check if the client participated in sum phase
+                --
+                -- Note: we cannot delete the sum_pk in the sum_dict because we
+                -- need the sum_dict later to delete the seed_dict
+                local sum_pk_exist = redis.call("HEXISTS", "sum_dict", sum_pk)
+                if sum_pk_exist == 0 then
+                    return -1
+                end
+
+                -- check if sum participated has not already submitted a mask
+                local mask_already_submitted = redis.call("SADD", "mask_submitted", sum_pk)
+                -- SADD returns 0 if the key already exists
+                if mask_already_submitted == 0 then
+                    return -2
+                end
+
+                redis.call("ZINCRBY", "mask_dict", 1, KEYS[1])
+
+                return 0
+            "#,
+        );
+
+        script
+            .key(MaskObjectWrite::from(mask))
+            .arg(PublicSigningKeyWrite::from(sum_pk))
+            .invoke_async(&mut self.connection)
             .await
     }
 
     /// Retrieves the two masks with the highest score.
```

Comment:
why *two* masks?

Reply:
In the state unmask, we check whether the two masks with the highest score do not have the same score.